arch/risc-v/include/mode.h

```diff
@@ -0,0 +1,103 @@
+/****************************************************************************
+ * arch/risc-v/include/mode.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_RISCV_INCLUDE_MODE_H
+#define __ARCH_RISCV_INCLUDE_MODE_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <arch/csr.h>
+#include <arch/irq.h>
```

Comment:
According to comment `csr.h` should not be included directly, but only indirectly. I address this in https://github.com/apache/incubator-nuttx/pull/5917 so this place should also be updated.

Reply:
#include <arch/arch.h> instead ?

Reply:
```
/* This file should never be included directly but, rather, only indirectly
 * through nuttx/irq.h
 */
```

Reply:
At least one issue that immediately arises from this, is that the assembly files start crying about missing CSR definitions. They cannot be included via nuttx.irq because

```
#ifndef __ASSEMBLY__
#include <stdint.h>
#include <nuttx/irq.h>
#include <arch/csr.h>
#include <arch/chip/irq.h>
#include <arch/mode.h>
#endif
```

So what do you suggest ?

Reply:
I'll move `#include <arch/csr.h>` out of `#ifndef __ASSEMBLY__` in https://github.com/apache/incubator-nuttx/pull/5917

Reply:
Probably `#include <arch/chip/irq.h>` also can be moved. I'll check

Reply:
@pkarashchenko I ran into bit of an issue with the macros you introduced into riscv_internal.h. Some of the macros are needed in header files that live in the /include folder, so the scope must be widened.

I will move them to arch.h which is quite ubiquitous , we can move them elsewhere or to a new file if you think they should be in a more private place.

Reply:
@pkarashchenko please check the commit that moves the macros.

Reply:
Initially I thought to place macro in `arch/risc-v/include/arch.h` but didn't found any strong reason to do that in my initial change.

Reply:
> Initially I thought to place macro in `arch/risc-v/include/arch.h` but didn't found any strong reason to do that in my initial change.

I agree that arch.h is the logical place, but it went to a circular inclusion hell I was not interested in solving -> I made the new file.

Reply:
https://github.com/apache/incubator-nuttx/pull/5922 please check this. I will try to checkout your branch and rebase on top of these changes.

Reply:
> #5922 please check this. I will try to checkout your branch and rebase on top of these changes.

Thanks, I'll check it out. You'll probably have to do a bit of hacking to get the SW to build, as a lot of code is still missing from the "final working solution".